firmware/application/freqman_db.cpp

```diff
@@ -0,0 +1,264 @@
+// TODO: parse_int seems to hang on invalid input.
+bool parse_freqman_entry(std::string_view str, freqman_entry& entry) {
+    if (str.empty() || str[0] == '#')
+        return false;
+
+    auto cols = split_string(str, ',');
+    entry = freqman_entry{};
+
+    for (auto col : cols) {
+        if (col.empty())
+            continue;
+
+        auto pair = split_string(col, '=');
+        if (pair.size() != 2)
+            continue;
+
+        auto key = pair[0];
+        auto value = pair[1];
+
+        if (key == "a") {
+            entry.type = freqman_type::Range;
+            parse_int(value, entry.frequency_a);
+        } else if (key == "b") {
+            parse_int(value, entry.frequency_b);
+        } else if (key == "bw") {
+            // NB: Requires modulation to be set first
+            if (entry.modulation < std::size(freqman_bandwidths)) {
+                entry.bandwidth = find_by_name(freqman_bandwidths[entry.modulation], value);
+            }
+        } else if (key == "c") {
+            // Split into whole and fractional parts.
+            auto parts = split_string(value, '.');
+            int32_t tone_freq = 0;
+            int32_t whole_part = 0;
+            parse_int(parts[0], whole_part);
+
+            // Tones are stored as frequency / 100 for some reason.
+            // E.g. 14572 would be 145.7 (NB: 1s place is dropped).
+            // TODO: Might be easier to just store the codes?
+            // Multiply the whole part by 100 to get the tone frequency.
+            tone_freq = whole_part * 100;
+
+            // Add the fractional part, if present.
+            if (parts.size() > 1) {
+                auto c = parts[1].front();
+                auto digit = std::isdigit(c) ? c - '0' : 0;
+                tone_freq += digit * 10;
+            }
+            entry.tone = static_cast<freqman_index_t>(
+                tonekey::tone_key_index_by_value(tone_freq));
+        } else if (key == "d") {
+            entry.description = trim(value);
+        } else if (key == "f") {
+            entry.type = freqman_type::Single;
+            parse_int(value, entry.frequency_a);
+        } else if (key == "m") {
+            entry.modulation = find_by_name(freqman_modulations, value);
+        } else if (key == "r") {
+            entry.type = freqman_type::HamRadio;
+            parse_int(value, entry.frequency_a);
+        } else if (key == "s") {
+            entry.step = find_by_name(freqman_steps_short, value);
+        } else if (key == "t") {
+            parse_int(value, entry.frequency_b);
+        }
+    }
+
+    return true;
+}
+
+bool parse_freqman_file(const fs::path& path, freqman_db& db, freqman_load_options options) {
+    File f;
+    auto error = f.open(path);
+    if (error)
+        return false;
+
+    auto reader = FileLineReader(f);
+    auto line_count = count_lines(reader);
+
+    // Attempt to avoid a re-alloc if possible.
+    db.clear();
+    db.reserve(line_count);
```

Comment:
Just read on it right now after a google search. 
For the reference, reserving reduce memory reallocation and copies.
A good explanation here: https://www.acodersjourney.com/6-tips-supercharge-cpp-11-vector-performance/